TypeCobol/Compiler/Diagnostics/CrossChecker.cs

```diff
@@ -240,86 +240,85 @@ public override bool Visit(DataDefinition dataDefinition)
-            //Level 88 and 66 cannot have Children.
-            if (levelNumber != null && (levelNumber.Value == 88 || levelNumber.Value == 66) && dataDefinition.ChildrenCount != 0)
+            if (levelNumber != null)
             {
-                DiagnosticUtils.AddError(dataDefinition.CodeElement,
-                    "The variable '" + dataDefinition.Name + "' with level 88 and 66 cannot be group item.");
-            }
-            if (dataDefinition.Usage != null)
-            {
-                if (levelNumber != null && (levelNumber.Value == 88 || levelNumber.Value == 66))
-                {//page 229 : Level 88 and 66 cannot have use
-                    DiagnosticUtils.AddError(dataDefinition.CodeElement,
-                        "The variable '" + dataDefinition.Name + "' with level 88 and 66 cannot have USAGE.");
+                if (dataDefinitionParent != null)
+                {
+                    if (dataDefinitionParent.DataType == DataType.Boolean && levelNumber.Value == 88)
+                    {
+                        DiagnosticUtils.AddError(dataDefinition.CodeElement,
+                            "The Level 88 symbol '" + dataDefinition.Name + "' cannot be declared under a BOOL typed symbol");
+                    }
+                }
+                else
+                {
+                    if (!(levelNumber.Value == 01 || levelNumber.Value == 77))
+                    {
+                        DiagnosticUtils.AddError(dataDefinition.CodeElement,
+                            "The variable '" + dataDefinition.Name + "' can only be of level 01 or 77");
+                    }
+                    
+                }
+                //Level 88 and 66 cannot have Children.
+                if ((levelNumber.Value == 88 || levelNumber.Value == 66))
+                {
+                    if (dataDefinition.ChildrenCount != 0)
+                    {
+                        DiagnosticUtils.AddError(dataDefinition.CodeElement,
+                            "The variable '" + dataDefinition.Name + "' with level 88 and 66 cannot be group item.");
+                    }
+
+                    if (dataDefinition.Usage != null)
+                    {
+                        DiagnosticUtils.AddError(dataDefinition.CodeElement,
+                            "The variable '" + dataDefinition.Name + "' with level 88 and 66 cannot have USAGE.");
+                    }
                 }
+
             }
+
             if (dataDefinition.Picture != null)
             {//only children with level 77 or 88 can be children of a PICTURE Elementary Item
-                if (dataDefinition.Children.Any(elem => elem.CodeElement != null && 
-                                                        elem.CodeElement.Type != CodeElementType.DataConditionEntry &&
-                                                        elem.CodeElement.Type != CodeElementType.DataRenamesEntry))
+                if (HasOtherChildrenThanAllowed(dataDefinition))
```

Comment:
For performance reason, place only call HasOtherChildrenThanAllowed once in this test.